R/SEIRDAge.R

```diff
@@ -0,0 +1,333 @@
+#' \deqn{\frac{dI_i(t)}{dt} = k E_i(t) - g I_i(t)}
+#' \deqn{\frac{dR_i(t)}{dt} = g I_i(t)}
+#' \deqn{\frac{dD_i(t)}{dt} = \mu I_i(t)}
+
+#' where C is a contact matrix whose elements represents the
+#' contact between different age groups (rows) with age groups of
+#' people they come in contact with (columns). This function relies on the 
+#' package deSolve to numerically integrate the set of equations above.
+#' 
+#'
+#' @param object An object of the class SEIRDAge.
+#' @param times (vector) time sequence over which to solve the model.
+#'        Must be of the form seq(t_start,t_end,by=t_step). Default time series
+#'        is seq(0, 100, by = 1).
+#' @param solve_method A string indicating the chosen numerical integration
+#' method for solving the ode system. Default is `lsoda` which is also the
+#' default for the ode function in the deSolve package used in this function.
+#'
+#' @return data frame containing the time vector and time series of S, R, I and
+#' D population fractions for each age group outputs with incidence numbers
+#' for each age group.
+#' @rdname SEIRDAge-class
+#' @aliases run,ANY,ANY-method
+#' @importFrom magrittr "%>%"
+#' @importFrom rlang .data
+#' @export
+setMethod(
+  "run", 'SEIRDAge',
+  function(object, times, solve_method = 'lsoda') {
+
+    # error if times is not a vector or list of doubles
+    if(!is.double(times)){
+      stop('Evaluation times of the model storage format must be a vector.')
+    }
+    if(!is.numeric(times)){
+      stop('Evaluation times of the model storage format must numeric.')
+    }
+    if(!(all(diff(times) > 0))){
+      stop('Evaluation times of the model storage format must be increasing')
+    }
+    
+    #fetch number of age catagories
+    age <- object@n_age_categories
+
+    # set initial state vector
+    state <- c(S = initial_conditions(object)$S0,
+               E = initial_conditions(object)$E0,
+               I = initial_conditions(object)$I0,
+               R = initial_conditions(object)$R0,
+               D = initial_conditions(object)$D0)
+
+    # set parameters vector
+    parameters <- c(b = transmission_parameters(object)$b,
+                    k = transmission_parameters(object)$k,
+                    g = transmission_parameters(object)$g,
+                    mu = transmission_parameters(object)$mu)
+    
+    # fetch contact matrix of the instance
+    C = object@contact_matrix
+
+    # function for RHS of ode system
+    right_hand_side <- function(t, state, parameters) {
+      with(
+        as.list(c(state, parameters)),
+        {
+          S <- state[1:age]
+          E <- state[(age + 1):(2 * age)]
+          I <- state[(2 * age + 1):(3 * age)]
+          R <- state[(3 * age + 1):(4 * age)]
+          D <- state[(4 * age + 1):(5 * age)]
+          
+          
+          # rate of change
+          dS <- -b * S * C %*% I
+          dE <- b * S * C %*% I - k * E
+          dI <- k * E - g * I
```

Comment:
```suggestion
          dI <- k * E - g * I - mu * I
```
Do we need to account for I -> D? (also in the documentation above)

Reply:
You are right! the RHS of sum of the equations of the system must be 0 to keep the total population constant in time